Apps/Inc/fifo_protected.h

```diff
@@ -0,0 +1,337 @@
+                0,
+                &err);
+    assertOSError(OS_NONE_LOC, err);
+
+    return fifo;
+}
+
+/**
+ * @brief Initialize a fifo by reference.
+ * 
+ * This is going to be faster than _new().
+ * This does not erase all of the contents of the fifo, but
+ * rather marks it as "empty".
+ */
+static inline void __attribute__((unused))
+RENEW (FIFO_TYPE_NAME * fifo) {
+    OS_ERR err;
+    CPU_TS ticks;
+
+    // Create mutex and semaphore if not already created
+    OSMutexCreate(&(fifo->FIFO_MUTEX_NAME), XSTR((FIFO_MUTEX_NAME)), &err);
+    
+    // Don't error out if the mutex already exists, just move on.
+    if(err != OS_ERR_OBJ_CREATED)
+        assertOSError(OS_NONE_LOC, err);
+
+    OSSemCreate(&(fifo->FIFO_SEM4_NAME),
+                XSTR((FIFO_SEM4_NAME)),
+                0,
+                &err);
+    
+    // Don't error out if the semaphore already exists, just move on.
+    if(err != OS_ERR_OBJ_CREATED)
+        assertOSError(OS_NONE_LOC, err);
+    
+    // Renew
+    OSSemSet(&(fifo->FIFO_SEM4_NAME), 0, &err);
+    assertOSError(OS_NONE_LOC, err);
+
+    OSMutexPend(&(fifo->FIFO_MUTEX_NAME), 0, OS_OPT_PEND_BLOCKING, &ticks, &err);
+    assertOSError(OS_NONE_LOC, err);
+    
+    if(fifo != NULL) {
+        fifo->get = fifo->put;
+    }
+
+    OSMutexPost(&(fifo->FIFO_MUTEX_NAME), OS_OPT_POST_NONE, &err);
+    assertOSError(OS_NONE_LOC, err);
+}
+
+/**
+ * @brief Determine whether the fifo is empty.
+ * 
+ * If the type of the fifo is myfifo_t, then this function
+ * will be called myfifo_is_empty().
+ * 
+ * @param fifo A pointer to the fifo
+ * @return true If empty
+ * @return false If not empty
+ */
+static bool
+IS_EMPTY (FIFO_TYPE_NAME *fifo) {
+    return fifo->put == fifo->get;
+}
+
+/**
+ * @brief Determine whether the fifo is full.
+ * 
+ * If the type of the fifo is myfifo_t, then this function
+ * will be called myfifio_is_full().
+ * 
+ * @param fifo A pointer to the fifo
+ * @return true If full
+ * @return false If not full
+ */
+static bool
+IS_FULL (FIFO_TYPE_NAME *fifo) {
+    return (fifo->put + 1) % FIFO_SIZE == fifo->get;
+}
+
+/**
+ * @brief Get the next element from the fifo
+ * 
+ * If the type of the fifo is myfifo_t, then this function
+ * will be called myfifo_get().
+ * 
+ * @param fifo A pointer to the fifo
+ * @param elem A pointer to an element to use for storage
+ * @return true if successful
+ * @return false if unsuccessful
+ */
+static bool __attribute__((unused))
+GET (FIFO_TYPE_NAME *fifo, FIFO_TYPE *elem) {
+    OS_ERR err;
+    CPU_TS ticks;
+    bool success = false;
+    
+    OSSemPend(&(fifo->FIFO_SEM4_NAME), 0, OS_OPT_PEND_BLOCKING, &ticks, &err);
+    assertOSError(OS_NONE_LOC, err);
+
+    OSMutexPend(&(fifo->FIFO_MUTEX_NAME), 0, OS_OPT_PEND_BLOCKING, &ticks, &err);
+    assertOSError(OS_NONE_LOC, err);
+    
+    if(!IS_EMPTY(fifo)) {
+        *elem = fifo->buffer[fifo->get];
+        fifo->get = (fifo->get + 1) % FIFO_SIZE;
+        success = true;
+    }
```

Comment:
Do we need to check if full. Shouldn't the semaphore keep track of the resources inside?